### PR TITLE
[RemoveDIs][DebugInfo] Enable creation of DPVAssigns, update outstanding AT tests

### PR DESCRIPTION
**Pull request by SLTozer**
This is the final patch for DPVAssign support, implementing the actual creation of DPVAssigns and allowing them to be converted along with dbg.values and dbg.declares. Numerous tests landed in previous patches will no longer be rotten after this patch lands (previously they would trivially pass due to DPVAssigns not actually being used), and a further batch of tests have been added here that require the changes in this patch before they pass.

### CODE REVIEW

File: llvm/lib/IR/DebugInfo.cpp [hunk 1]
```diff
@@ -2206,32 +2217,40 @@ bool AssignmentTrackingPass::runOnFunction(Function &F) {
...
-        // FIXME: Skip VLAs for now (let these variables use dbg.declares).
-        if (!Alloca->isStaticAlloca())
-          continue;
-        // Similarly, skip scalable vectors (use dbg.declares instead).
-        if (auto Sz = Alloca->getAllocationSize(*DL); Sz && Sz->isScalable())
-          continue;
-        DbgDeclares[Alloca].insert(DDI);
-        Vars[Alloca].insert(VarRecord(DDI));
-      }
+      ProcessDeclare(DDI, DbgDeclares);
```
Comment (OCHyams) [inline llvm/lib/IR/DebugInfo.cpp:2253]: nit, I think this should be simplified to:
```
      if (DbgDeclareInst *DDI = dyn_cast<DbgDeclareInst>(&I))
        ProcessDeclare(DDI, DbgDeclares);
```

File: llvm/test/DebugInfo/Generic/assignment-tracking/slp-vectorizer/merge-scalars.ll [hunk 2]
```diff
@@ -60,6 +60,7 @@ entry:
   %arrayidx10 = getelementptr inbounds [4 x float], ptr %quad, i64 0, i64 3, !dbg !58
   store float %add9, ptr %arrayidx10, align 4, !dbg !59, !DIAssignID !60
   call void @llvm.dbg.assign(metadata float %add9, metadata !16, metadata !DIExpression(DW_OP_LLVM_fragment, 96, 32), metadata !60, metadata ptr %arrayidx10, metadata !DIExpression()), !dbg !23
+  call void @llvm.dbg.value(metadata float %add9, metadata !16, metadata !DIExpression(DW_OP_LLVM_fragment, 96, 32)), !dbg !23
```
Comment (OCHyams) [inline llvm/test/DebugInfo/Generic/assignment-tracking/slp-vectorizer/merge-scalars.ll:63]: where does this change come from?

**Review by OCHyams — COMMENTED** [review]
Couple of inline comments, otherwise LGTM

File: llvm/test/DebugInfo/Generic/assignment-tracking/slp-vectorizer/merge-scalars.ll (hunk 2 above)
Comment (SLTozer) [reply llvm/test/DebugInfo/Generic/assignment-tracking/slp-vectorizer/merge-scalars.ll:63]: Think this is an accidental WIP/testing change that somehow got in, good catch.

**Review by OCHyams — APPROVED** [review]

**Pull request merged**